src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs

```diff
@@ -356,6 +373,7 @@ private BsonDocument CreateChangeStreamStage()
             {
                 { "fullDocument", ToString(_fullDocument) },
                 { "allChangesForCluster", true, _collectionNamespace == null && _databaseNamespace == null },
+                { "startAfter", _startAfter, _startAfter != null},
```

Comment:
CreateChangeStreamStage should be refactored to consider whether it is a resuming operation and to use the "effective" resume/start values:

    private BsonDocument CreateChangeStreamStage(bool resuming)
    {
        var resumeStartValues = GetEffectiveResumeStartValues(resuming);

        var changeStreamOptions = new BsonDocument
        {
            { "fullDocument", ToString(_fullDocument) },
            { "allChangesForCluster", true, _collectionNamespace == null && _databaseNamespace == null },
            { "startAfter", resumeStartValues.StartAfter, resumeStartValues.StartAfter != null},
            { "startAtOperationTime", resumeStartValues.StartAtOperationTime, resumeStartValues.StartAtOperationTime != null },
            { "resumeAfter", resumeStartValues.ResumeAfter, resumeStartValues.ResumeAfter != null }
        };
        return new BsonDocument("$changeStream", changeStreamOptions);
    }

It is this approach of "effective resume/start values" that allows us to NOT overwrite the original values supplied by the creator of the operation.

Reply:
The GetEffectiveResumeStartValues should look like:

    private ResumeStartValue GetEffectiveResumeStartValues(bool resuming)
    {
        if (resuming)
        {
            if (_documentResumeToken != null)
            {
                return new ResumeStartValue { ResumeAfter = _documentResumeToken };
            }

            if (_startAfter != null)
            {
                return new ResumeStartValue { ResumeAfter = _startAfter };
            }

            if (_resumeAfter != null)
            {
                return new ResumeStartValue { ResumeAfter = _resumeAfter };
            }

            if (_startAtOperationTime != null || _initialOperationTime != null)
            {
                return new ResumeStartValue { StartAtOperationTime = _startAtOperationTime ?? _initialOperationTime };
            }
        }

        return new ResumeStartValue { ResumeAfter = _resumeAfter, StartAfter = _startAfter, StartAtOperationTime = _startAtOperationTime };
    }

NOTE: by refactoring to not change the original _startAfter etc.. values supplied by the creator of the operation this code now reads almost exactly like the "Resume Process" section of the spec (minus the postBatchResumeToken which will be added later).

The return value is a simple DTO:

    private struct ResumeStartValue
    {
        public BsonDocument ResumeAfter { get; set; }
        public BsonDocument StartAfter { get; set; }
        public BsonTimestamp StartAtOperationTime { get; set; }
    }

Reply:
Done